promql/parser/lex.go

```diff
@@ -901,19 +906,78 @@ func acceptRemainingDuration(l *Lexer) bool {
 // scanNumber scans numbers of different formats. The scanned Item is
 // not necessarily a valid number. This case is caught by the parser.
 func (l *Lexer) scanNumber() bool {
-	digits := "0123456789"
+	// Modify the digit pattern if the number is hexadecimal.
+	digitPattern := "0123456789"
 	// Disallow hexadecimal in series descriptions as the syntax is ambiguous.
-	if !l.seriesDesc && l.accept("0") && l.accept("xX") {
-		digits = "0123456789abcdefABCDEF"
+	if !l.seriesDesc &&
+		l.accept("0") && l.accept("xX") {
+		l.accept("_") // eg., 0X_1FFFP-16 == 0.1249847412109375
+		digitPattern = "0123456789abcdefABCDEF"
 	}
-	l.acceptRun(digits)
-	if l.accept(".") {
-		l.acceptRun(digits)
-	}
-	if l.accept("eE") {
-		l.accept("+-")
-		l.acceptRun("0123456789")
+	// Define dot, exponent, and underscore patterns.
+	dotPattern := "."
+	exponentPattern := "eE"
+	underscorePattern := "_"
+	// Anti-patterns are rune sets that cannot follow their respective rune.
+	dotAntiPattern := "_."
+	exponentAntiPattern := "._eE"   // and EOL.
+	underscoreAntiPattern := "._eE" // and EOL.
```

Comment:
These are not modified like `digitPattern` is. Let's move them to the package scope as `const` values.

Reply:
+1 on the `const` change, but I'm not sure about changing the scope for these (since they are only used in this method)? LMK if you think otherwise and I'll make the change.